docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,403 @@
+- The eight value is **Argument[0].Paramter[0]**, which is the access path the first parameter of the **Function** argument of **map** (the lambda parameter **e** in the example).
+
+For the second row the
+
+- The seventh value is **Argument[0].ReturnValue**, which is the access path to the return value of the **Function** argument of **map** (the return value of the lambda in the example).
+- The eighth value is **ReturnValue.Element**, which is the access path to the elements of the return value of **map** (the elements of the stream **l** in the example).
+
+The remaining values for both rows
+
+- The ninth value **value** is the kind of the flow. **value** means that the value is preserved.
+- The tenth value **manual** is the provenance of the summary, which is used to identify the origin of the summary.
+
+That is, the first row models that there is value flow from the elements of the qualifier stream into the first argument of the Function provided to **map** and the second row models that there is value flow from the return value of the Function to the elements of the stream returned from **map**.
+
+Example: Add a **neutral** method.
+----------------------------------
+In this example we will see, how to define the **now** method as being neutral.
+This is purely for consistency and has no impact on the analysis.
+A neutral model is used to define that there is no flow through a method.
+Please note that the neutral model for the **now** method is already added.
+
+.. code-block:: java
+   
+   public static Instant taintflow() {
+       Instant t = Instant.now(); // There is no flow from now to t.
+       return t;
+   }
+
+.. code-block:: yaml
+
+   extensions:
+   - addsTo:
+       pack: codeql/java-all
+       extensible: neutralModel
+     data:
+       - ["java.time", "Instant", "now", "()", "manual"]
+
+Reasoning:
+
+Since we are adding a neutral model, we need to add tuples to the **neutralModel** extension point.
+The first four values are used to identify the method (callable) which we are defining as a neutral and the fifth value is the provenance (origin) of the neutral.
+
+- The first value **java.time** is the package name.
+- The second value **Instant** is the class (type) name.
+- The third value **now** is the method name.
+- The fourth value **()** is the method input type signature.
+- The fifth value **manual** is the provenance of the neutral.
+
+Reference material
+------------------
+
+The following sections provide reference material for extension points.
+This includes descriptions of each of the arguments (eg. access paths, kinds and provenance).
+
+Extension points
+----------------
+
+Below is a description of the columns for each extension point.
+Sources, Sinks, Summaries and Neutrals are commonly known as Models.
+The semantics of many of the columns of the extension points are shared.
+
+The shared columns are:
+
+- **package**: Name of the package.
+- **type**: Name of the type.
```

Comment:
```suggestion
- **type**: Name of the type containing the element to be modeled.
```

(This is necessary to distinguish from other types involved with the modeled element, like return type, etc).